src/crimson/os/seastore/onode_manager/staged-fltree/node.cc

```diff
@@ -0,0 +1,805 @@
+      return do_get_tree_stats(c, stats).safe_then([&stats] {
+        return stats;
+      });
+    }
+  );
+}
+
+std::ostream& Node::dump(std::ostream& os) const {
+  return impl->dump(os);
+}
+
+std::ostream& Node::dump_brief(std::ostream& os) const {
+  return impl->dump_brief(os);
+}
+
+void Node::test_make_destructable(
+    context_t c, NodeExtentMutable& mut, Super::URef&& _super) {
+  impl->test_set_tail(mut);
+  make_root(c, std::move(_super));
+}
+
+node_future<> Node::mkfs(context_t c, RootNodeTracker& root_tracker) {
+  return LeafNode::allocate_root(c, root_tracker
+  ).safe_then([](auto ret) { /* FIXME: discard_result(); */ });
+}
+
+node_future<Ref<Node>> Node::load_root(context_t c, RootNodeTracker& root_tracker) {
+  return c.nm.get_super(c.t, root_tracker
+  ).safe_then([c, &root_tracker](auto&& _super) {
+    auto root_addr = _super->get_root_laddr();
+    assert(root_addr != L_ADDR_NULL);
+    return Node::load(c, root_addr, true
+    ).safe_then([c, _super = std::move(_super),
+                 &root_tracker](auto root) mutable {
+      assert(root->impl->field_type() == field_type_t::N0);
+      root->as_root(std::move(_super));
+      assert(root == root_tracker.get_root(c.t));
+      return node_ertr::make_ready_future<Ref<Node>>(root);
+    });
+  });
+}
+
+void Node::make_root(context_t c, Super::URef&& _super) {
+  _super->write_root_laddr(c, impl->laddr());
+  as_root(std::move(_super));
+}
+
+void Node::as_root(Super::URef&& _super) {
+  assert(!super && !_parent_info);
+  assert(_super->get_root_laddr() == impl->laddr());
+  assert(impl->is_level_tail());
+  super = std::move(_super);
+  super->do_track_root(*this);
+}
+
+node_future<> Node::upgrade_root(context_t c) {
+  assert(is_root());
+  assert(impl->is_level_tail());
+  assert(impl->field_type() == field_type_t::N0);
+  super->do_untrack_root(*this);
+  return InternalNode::allocate_root(c, impl->level(), impl->laddr(), std::move(super)
+  ).safe_then([this](auto new_root) {
+    as_child(search_position_t::end(), new_root);
+  });
+}
+
+template <bool VALIDATE>
+void Node::as_child(const search_position_t& pos, Ref<InternalNode> parent_node) {
+  assert(!super);
+  _parent_info = parent_info_t{pos, parent_node};
+  parent_info().ptr->do_track_child<VALIDATE>(*this);
+}
+template void Node::as_child<true>(const search_position_t&, Ref<InternalNode>);
+template void Node::as_child<false>(const search_position_t&, Ref<InternalNode>);
+
+node_future<> Node::insert_parent(context_t c, Ref<Node> right_node) {
+  assert(!is_root());
+  // TODO(cross-node string dedup)
+  return parent_info().ptr->apply_child_split(
+      c, parent_info().position, this, right_node);
+}
+
+node_future<Ref<Node>> Node::load(
+    context_t c, laddr_t addr, bool expect_is_level_tail) {
+  // NOTE:
+  // *option1: all types of node have the same length;
+  // option2: length is defined by node/field types;
+  // option3: length is totally flexible;
```

Comment:
Is the asterix meant to suggest that option1 is the one selected?  If so, it's not obvious.

Reply:
It's the current implemented option which is also the simplest. We can implement more flexible node size later if needed.

Reply:
Oh, that's totally fine, just clearer comment would help.

Reply:
fixed in https://github.com/ceph/ceph/pull/36986/commits/1c053f3425ac9bc2c3d88d2ffcd6d85d18aa7f9f